### PR TITLE
feat(contract): update Multicall to Multicall3

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Closes: #1576 

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Updated the changelog

### CODE REVIEW

**Comment by gakonst** [comment]
@mds1 @mattsse wdyt?

**Comment by mds1** [comment]
Supportive, I agree with all the rationale in #1576 which was the motivation for Multicall3

**Comment by DaniPopes** [comment]
Looks like we need to lock the entire `Multicall` module behind the `abigen` feature.. unless there's a way to generate the contract binding without the `derive`s?

**Comment by DaniPopes** [comment]
^ CC @mattsse @gakonst - new contract binding requires the derive macros from `ethers-contract-derive` which needs the `abigen` feature to be enabled

**Comment by gakonst** [comment]
Hmm should we just stop feature gating abigen as a feature altogether? @mattsse i always seem to need it

File: ethers-contract/src/multicall/mod.rs [hunk 1]
```diff
@@ -330,28 +569,91 @@ impl<M: Middleware> Multicall<M> {
...
+                                        "Illegal revert.\n{:?}\n{:?}",
+                                        call, res
+                                    )),
+                                ))
+                            }
+
+                            // Decode with "Error(string)" (0x08c379a0)
+                            if ret.len() >= 4 && ret[..4] == [0x08, 0xc3, 0x79, 0xa0] {
+                                decode(&[ParamType::String], &ret[4..])?.remove(0)
+                            } else if ret.is_empty() {
```
Comment (mattsse) [inline ethers-contract/src/multicall/mod.rs:630]: can use `String::decode(&ret[4..])` here
with `AbiDecode`

File: ethers-contract/src/multicall/multicall_contract.rs [hunk 2]
```diff
@@ -1,97 +1,911 @@
...
-    });
-    pub struct MulticallContract<M>(Contract<M>);
-    impl<M> Clone for MulticallContract<M> {
+    // Some macros like EthAbiType and EthAbiCodec expand into using "ethers_contract" which is not
+    // defined here
+    mod ethers_contract {
+        pub use crate::*;
+        // re-export derive macros to build this multicall in case abigen feature is not enabled
+        pub use ethers_contract_derive::*;
```
Comment (mattsse) [inline ethers-contract/src/multicall/multicall_contract.rs:12]: this reexport seems wrong

this reexports the derive macros not the traits

**Review by mattsse — CHANGES_REQUESTED** [review]
sorry, this one slipped through.

some nits, but lgtm in general

**Review by mattsse — APPROVED** [review]
docs + test lgtm

**Review by gakonst — APPROVED** [review]
LGTM - it does add an extra param on add_call but should be ok

**Pull request merged**